src/console.cpp

```diff
@@ -195,80 +204,36 @@ bool GetArgumentInteger(uint32 *value, const char *arg)
 }
 
 /**
- * Add an item to an alphabetically sorted list.
- * @param base first item of the list
- * @param item_new the item to add
+ * Creates a copy of a string with underscores removed from it
+ * @param name String to remove the underscores from.
+ * @return A copy of \a name, without underscores.
  */
-template<class T>
-void IConsoleAddSorted(T **base, T *item_new)
+static std::string RemoveUnderscores(const std::string &name)
 {
-	if (*base == nullptr) {
-		*base = item_new;
-		return;
-	}
-
-	T *item_before = nullptr;
-	T *item = *base;
-	/* The list is alphabetically sorted, insert the new item at the correct location */
-	while (item != nullptr) {
-		if (strcmp(item->name, item_new->name) > 0) break; // insert here
-
-		item_before = item;
-		item = item->next;
-	}
-
-	if (item_before == nullptr) {
-		*base = item_new;
-	} else {
-		item_before->next = item_new;
-	}
-
-	item_new->next = item;
-}
-
-/**
- * Remove underscores from a string; the string will be modified!
- * @param[in,out] name String to remove the underscores from.
- * @return \a name, with its contents modified.
- */
-char *RemoveUnderscores(char *name)
-{
-	char *q = name;
-	for (const char *p = name; *p != '\0'; p++) {
-		if (*p != '_') *q++ = *p;
-	}
-	*q = '\0';
-	return name;
+	std::string copy(name);
+	copy.erase(std::remove_if(copy.begin(), copy.end(), [](char c) { return c == '_'; }), copy.end());
+	return copy;
 }
 
 /**
  * Register a new command to be used in the console
  * @param name name of the command that will be used
  * @param proc function that will be called upon execution of command
  */
-void IConsoleCmdRegister(const char *name, IConsoleCmdProc *proc, IConsoleHook *hook)
+/* static */ void IConsole::CmdRegister(const std::string &name, IConsoleCmdProc *proc, IConsoleHook *hook)
 {
-	IConsoleCmd *item_new = MallocT<IConsoleCmd>(1);
-	item_new->name = RemoveUnderscores(stredup(name));
-	item_new->next = nullptr;
-	item_new->proc = proc;
-	item_new->hook = hook;
-
-	IConsoleAddSorted(&_iconsole_cmds, item_new);
+	IConsole::Commands().try_emplace(RemoveUnderscores(name), name, proc, hook);
```

Comment:
As suggested in [a PR to your fork](https://github.com/glx22/OpenTTD/pull/12/commits/6d175e3c81e96b80193d699949e7bb177bc1c1f8), this is a change of behaviour of the command registering mechanism, as commands' key in the map won't be the same as their name.
Could this have side-effects on actual command display/usage?

Reply:
Only on display (and for me it's an improvement). Usage is the same as before.

Reply:
I think this way of fixing what was fixed in d98ff7c4 is the more elegant way. Before that commit you were required to write the underscore in most commands and they were displayed with underscores. The commands are easier to read with underscores in them as it breaks the words up a bit. And functionally the change is mostly in what you see when the commands are listed.
Actually, I would say it even fixes the inconsistency between the naming of the commands in listcmds and other "documentation". Look for example at the (error) message you get for the help of the following commands: server_info, client_name, reset_company, start_ai, reload_ai, stop_ai, rescan_ai, rescan_game, rescan_newgrf, info_cmd, debug_level, list_cmds, list_aliases, say_company, say_client, company_pw, setting_newgame, list_settings, newgrf_profile and dump_info. They all show the underscore in the help message, so practically list_cmds should show the underscores as well.